// angular2/src/core/compiler/proto_view_factory.js
"use strict";
Object.defineProperties(module.exports, {
  ProtoViewFactory: {get: function() {
      return ProtoViewFactory;
    }},
  getChangeDetectorDefinitions: {get: function() {
      return getChangeDetectorDefinitions;
    }},
  __esModule: {value: true}
});
var $__angular2_47_src_47_di_47_annotations_95_impl__,
    $__angular2_47_src_47_facade_47_collection__,
    $__angular2_47_src_47_facade_47_lang__,
    $__angular2_47_src_47_reflection_47_reflection__,
    $__angular2_47_change_95_detection__,
    $__angular2_47_src_47_render_47_api__,
    $__view__,
    $__element_95_injector__;
var Injectable = ($__angular2_47_src_47_di_47_annotations_95_impl__ = require("angular2/src/di/annotations_impl"), $__angular2_47_src_47_di_47_annotations_95_impl__ && $__angular2_47_src_47_di_47_annotations_95_impl__.__esModule && $__angular2_47_src_47_di_47_annotations_95_impl__ || {default: $__angular2_47_src_47_di_47_annotations_95_impl__}).Injectable;
var $__1 = ($__angular2_47_src_47_facade_47_collection__ = require("angular2/src/facade/collection"), $__angular2_47_src_47_facade_47_collection__ && $__angular2_47_src_47_facade_47_collection__.__esModule && $__angular2_47_src_47_facade_47_collection__ || {default: $__angular2_47_src_47_facade_47_collection__}),
    List = $__1.List,
    ListWrapper = $__1.ListWrapper,
    MapWrapper = $__1.MapWrapper;
var $__2 = ($__angular2_47_src_47_facade_47_lang__ = require("angular2/src/facade/lang"), $__angular2_47_src_47_facade_47_lang__ && $__angular2_47_src_47_facade_47_lang__.__esModule && $__angular2_47_src_47_facade_47_lang__ || {default: $__angular2_47_src_47_facade_47_lang__}),
    isPresent = $__2.isPresent,
    isBlank = $__2.isBlank;
var reflector = ($__angular2_47_src_47_reflection_47_reflection__ = require("angular2/src/reflection/reflection"), $__angular2_47_src_47_reflection_47_reflection__ && $__angular2_47_src_47_reflection_47_reflection__.__esModule && $__angular2_47_src_47_reflection_47_reflection__ || {default: $__angular2_47_src_47_reflection_47_reflection__}).reflector;
var $__4 = ($__angular2_47_change_95_detection__ = require("angular2/change_detection"), $__angular2_47_change_95_detection__ && $__angular2_47_change_95_detection__.__esModule && $__angular2_47_change_95_detection__ || {default: $__angular2_47_change_95_detection__}),
    ChangeDetection = $__4.ChangeDetection,
    DirectiveIndex = $__4.DirectiveIndex,
    BindingRecord = $__4.BindingRecord,
    DirectiveRecord = $__4.DirectiveRecord,
    ProtoChangeDetector = $__4.ProtoChangeDetector,
    DEFAULT = $__4.DEFAULT,
    ChangeDetectorDefinition = $__4.ChangeDetectorDefinition;
var renderApi = ($__angular2_47_src_47_render_47_api__ = require("angular2/src/render/api"), $__angular2_47_src_47_render_47_api__ && $__angular2_47_src_47_render_47_api__.__esModule && $__angular2_47_src_47_render_47_api__ || {default: $__angular2_47_src_47_render_47_api__});
var AppProtoView = ($__view__ = require("./view"), $__view__ && $__view__.__esModule && $__view__ || {default: $__view__}).AppProtoView;
var $__6 = ($__element_95_injector__ = require("./element_injector"), $__element_95_injector__ && $__element_95_injector__.__esModule && $__element_95_injector__ || {default: $__element_95_injector__}),
    ProtoElementInjector = $__6.ProtoElementInjector,
    DirectiveBinding = $__6.DirectiveBinding;
var BindingRecordsCreator = function BindingRecordsCreator() {
  this._directiveRecordsMap = MapWrapper.create();
  this._textNodeIndex = 0;
};
($traceurRuntime.createClass)(BindingRecordsCreator, {
  getBindingRecords: function(elementBinders, allDirectiveMetadatas) {
    var bindings = [];
    for (var boundElementIndex = 0; boundElementIndex < elementBinders.length; boundElementIndex++) {
      var renderElementBinder = elementBinders[boundElementIndex];
      this._createTextNodeRecords(bindings, renderElementBinder);
      this._createElementPropertyRecords(bindings, boundElementIndex, renderElementBinder);
      this._createDirectiveRecords(bindings, boundElementIndex, renderElementBinder.directives, allDirectiveMetadatas);
    }
    return bindings;
  },
  getDirectiveRecords: function(elementBinders, allDirectiveMetadatas) {
    var directiveRecords = [];
    for (var elementIndex = 0; elementIndex < elementBinders.length; ++elementIndex) {
      var dirs = elementBinders[elementIndex].directives;
      for (var dirIndex = 0; dirIndex < dirs.length; ++dirIndex) {
        ListWrapper.push(directiveRecords, this._getDirectiveRecord(elementIndex, dirIndex, allDirectiveMetadatas[dirs[dirIndex].directiveIndex]));
      }
    }
    return directiveRecords;
  },
  _createTextNodeRecords: function(bindings, renderElementBinder) {
    var $__7 = this;
    if (isBlank(renderElementBinder.textBindings))
      return ;
    ListWrapper.forEach(renderElementBinder.textBindings, (function(b) {
      ListWrapper.push(bindings, BindingRecord.createForTextNode(b, $__7._textNodeIndex++));
    }));
  },
  _createElementPropertyRecords: function(bindings, boundElementIndex, renderElementBinder) {
    MapWrapper.forEach(renderElementBinder.propertyBindings, (function(astWithSource, propertyName) {
      ListWrapper.push(bindings, BindingRecord.createForElement(astWithSource, boundElementIndex, propertyName));
    }));
  },
  _createDirectiveRecords: function(bindings, boundElementIndex, directiveBinders, allDirectiveMetadatas) {
    var $__7 = this;
    for (var i = 0; i < directiveBinders.length; i++) {
      var directiveBinder = directiveBinders[i];
      var directiveMetadata = allDirectiveMetadatas[directiveBinder.directiveIndex];
      MapWrapper.forEach(directiveBinder.propertyBindings, (function(astWithSource, propertyName) {
        var setter = reflector.setter(propertyName);
        var directiveRecord = $__7._getDirectiveRecord(boundElementIndex, i, directiveMetadata);
        ListWrapper.push(bindings, BindingRecord.createForDirective(astWithSource, propertyName, setter, directiveRecord));
      }));
      MapWrapper.forEach(directiveBinder.hostPropertyBindings, (function(astWithSource, propertyName) {
        var dirIndex = new DirectiveIndex(boundElementIndex, i);
        ListWrapper.push(bindings, BindingRecord.createForHostProperty(dirIndex, astWithSource, propertyName));
      }));
    }
  },
  _getDirectiveRecord: function(boundElementIndex, directiveIndex, directiveMetadata) {
    var id = boundElementIndex * 100 + directiveIndex;
    if (!MapWrapper.contains(this._directiveRecordsMap, id)) {
      var changeDetection = directiveMetadata.changeDetection;
      MapWrapper.set(this._directiveRecordsMap, id, new DirectiveRecord(new DirectiveIndex(boundElementIndex, directiveIndex), directiveMetadata.callOnAllChangesDone, directiveMetadata.callOnChange, changeDetection));
    }
    return MapWrapper.get(this._directiveRecordsMap, id);
  }
}, {});
Object.defineProperty(BindingRecordsCreator.prototype.getBindingRecords, "parameters", {get: function() {
    return [[$traceurRuntime.genericType(List, renderApi.ElementBinder)], [$traceurRuntime.genericType(List, renderApi.DirectiveMetadata)]];
  }});
Object.defineProperty(BindingRecordsCreator.prototype.getDirectiveRecords, "parameters", {get: function() {
    return [[$traceurRuntime.genericType(List, renderApi.ElementBinder)], [$traceurRuntime.genericType(List, renderApi.DirectiveMetadata)]];
  }});
Object.defineProperty(BindingRecordsCreator.prototype._createTextNodeRecords, "parameters", {get: function() {
    return [[$traceurRuntime.genericType(List, BindingRecord)], [renderApi.ElementBinder]];
  }});
Object.defineProperty(BindingRecordsCreator.prototype._createElementPropertyRecords, "parameters", {get: function() {
    return [[$traceurRuntime.genericType(List, BindingRecord)], [$traceurRuntime.type.number], [renderApi.ElementBinder]];
  }});
Object.defineProperty(BindingRecordsCreator.prototype._createDirectiveRecords, "parameters", {get: function() {
    return [[$traceurRuntime.genericType(List, BindingRecord)], [$traceurRuntime.type.number], [$traceurRuntime.genericType(List, renderApi.DirectiveBinder)], [$traceurRuntime.genericType(List, renderApi.DirectiveMetadata)]];
  }});
Object.defineProperty(BindingRecordsCreator.prototype._getDirectiveRecord, "parameters", {get: function() {
    return [[$traceurRuntime.type.number], [$traceurRuntime.type.number], [renderApi.DirectiveMetadata]];
  }});
var ProtoViewFactory = function ProtoViewFactory(changeDetection) {
  this._changeDetection = changeDetection;
};
($traceurRuntime.createClass)(ProtoViewFactory, {createAppProtoViews: function(hostComponentBinding, rootRenderProtoView, allDirectives) {
    var $__7 = this;
    var allRenderDirectiveMetadata = ListWrapper.map(allDirectives, (function(directiveBinding) {
      return directiveBinding.metadata;
    }));
    var nestedPvsWithIndex = _collectNestedProtoViews(rootRenderProtoView);
    var nestedPvVariableBindings = _collectNestedProtoViewsVariableBindings(nestedPvsWithIndex);
    var nestedPvVariableNames = _collectNestedProtoViewsVariableNames(nestedPvsWithIndex, nestedPvVariableBindings);
    var changeDetectorDefs = _getChangeDetectorDefinitions(hostComponentBinding.metadata, nestedPvsWithIndex, nestedPvVariableNames, allRenderDirectiveMetadata);
    var protoChangeDetectors = ListWrapper.map(changeDetectorDefs, (function(changeDetectorDef) {
      return $__7._changeDetection.createProtoChangeDetector(changeDetectorDef);
    }));
    var appProtoViews = ListWrapper.createFixedSize(nestedPvsWithIndex.length);
    ListWrapper.forEach(nestedPvsWithIndex, (function(pvWithIndex) {
      var appProtoView = _createAppProtoView(pvWithIndex.renderProtoView, protoChangeDetectors[pvWithIndex.index], nestedPvVariableBindings[pvWithIndex.index], allDirectives);
      if (isPresent(pvWithIndex.parentIndex)) {
        var parentView = appProtoViews[pvWithIndex.parentIndex];
        parentView.elementBinders[pvWithIndex.boundElementIndex].nestedProtoView = appProtoView;
      }
      appProtoViews[pvWithIndex.index] = appProtoView;
    }));
    return appProtoViews;
  }}, {});
Object.defineProperty(ProtoViewFactory, "annotations", {get: function() {
    return [new Injectable()];
  }});
Object.defineProperty(ProtoViewFactory, "parameters", {get: function() {
    return [[ChangeDetection]];
  }});
Object.defineProperty(ProtoViewFactory.prototype.createAppProtoViews, "parameters", {get: function() {
    return [[DirectiveBinding], [renderApi.ProtoViewDto], [$traceurRuntime.genericType(List, DirectiveBinding)]];
  }});
function getChangeDetectorDefinitions(hostComponentMetadata, rootRenderProtoView, allRenderDirectiveMetadata) {
  var nestedPvsWithIndex = _collectNestedProtoViews(rootRenderProtoView);
  var nestedPvVariableBindings = _collectNestedProtoViewsVariableBindings(nestedPvsWithIndex);
  var nestedPvVariableNames = _collectNestedProtoViewsVariableNames(nestedPvsWithIndex, nestedPvVariableBindings);
  return _getChangeDetectorDefinitions(hostComponentMetadata, nestedPvsWithIndex, nestedPvVariableNames, allRenderDirectiveMetadata);
}
Object.defineProperty(getChangeDetectorDefinitions, "parameters", {get: function() {
    return [[renderApi.DirectiveMetadata], [renderApi.ProtoViewDto], [$traceurRuntime.genericType(List, renderApi.DirectiveMetadata)]];
  }});
function _collectNestedProtoViews(renderProtoView) {
  var parentIndex = arguments[1] !== (void 0) ? arguments[1] : null;
  var boundElementIndex = arguments[2] !== (void 0) ? arguments[2] : null;
  var result = arguments[3] !== (void 0) ? arguments[3] : null;
  if (isBlank(result)) {
    result = [];
  }
  ListWrapper.push(result, new RenderProtoViewWithIndex(renderProtoView, result.length, parentIndex, boundElementIndex));
  var currentIndex = result.length - 1;
  var childBoundElementIndex = 0;
  ListWrapper.forEach(renderProtoView.elementBinders, (function(elementBinder) {
    if (isPresent(elementBinder.nestedProtoView)) {
      _collectNestedProtoViews(elementBinder.nestedProtoView, currentIndex, childBoundElementIndex, result);
    }
    childBoundElementIndex++;
  }));
  return result;
}
Object.defineProperty(_collectNestedProtoViews, "parameters", {get: function() {
    return [[renderApi.ProtoViewDto], [$traceurRuntime.type.number], [], [$traceurRuntime.genericType(List, RenderProtoViewWithIndex)]];
  }});
function _getChangeDetectorDefinitions(hostComponentMetadata, nestedPvsWithIndex, nestedPvVariableNames, allRenderDirectiveMetadata) {
  return ListWrapper.map(nestedPvsWithIndex, (function(pvWithIndex) {
    var elementBinders = pvWithIndex.renderProtoView.elementBinders;
    var bindingRecordsCreator = new BindingRecordsCreator();
    var bindingRecords = bindingRecordsCreator.getBindingRecords(elementBinders, allRenderDirectiveMetadata);
    var directiveRecords = bindingRecordsCreator.getDirectiveRecords(elementBinders, allRenderDirectiveMetadata);
    var strategyName = DEFAULT;
    var typeString;
    if (pvWithIndex.renderProtoView.type === renderApi.ProtoViewDto.COMPONENT_VIEW_TYPE) {
      strategyName = hostComponentMetadata.changeDetection;
      typeString = 'comp';
    } else if (pvWithIndex.renderProtoView.type === renderApi.ProtoViewDto.HOST_VIEW_TYPE) {
      typeString = 'host';
    } else {
      typeString = 'embedded';
    }
    var id = (hostComponentMetadata.id + "_" + typeString + "_" + pvWithIndex.index);
    var variableNames = nestedPvVariableNames[pvWithIndex.index];
    return new ChangeDetectorDefinition(id, strategyName, variableNames, bindingRecords, directiveRecords);
  }));
}
Object.defineProperty(_getChangeDetectorDefinitions, "parameters", {get: function() {
    return [[renderApi.DirectiveMetadata], [$traceurRuntime.genericType(List, RenderProtoViewWithIndex)], [$traceurRuntime.genericType(List, $traceurRuntime.genericType(List, $traceurRuntime.type.string))], [$traceurRuntime.genericType(List, renderApi.DirectiveMetadata)]];
  }});
function _createAppProtoView(renderProtoView, protoChangeDetector, variableBindings, allDirectives) {
  var elementBinders = renderProtoView.elementBinders;
  var protoView = new AppProtoView(renderProtoView.render, protoChangeDetector, variableBindings);
  _createElementBinders(protoView, elementBinders, allDirectives);
  _bindDirectiveEvents(protoView, elementBinders);
  return protoView;
}
Object.defineProperty(_createAppProtoView, "parameters", {get: function() {
    return [[renderApi.ProtoViewDto], [ProtoChangeDetector], [$traceurRuntime.genericType(Map, $traceurRuntime.type.string, $traceurRuntime.type.string)], [$traceurRuntime.genericType(List, DirectiveBinding)]];
  }});
function _collectNestedProtoViewsVariableBindings(nestedPvsWithIndex) {
  return ListWrapper.map(nestedPvsWithIndex, (function(pvWithIndex) {
    return _createVariableBindings(pvWithIndex.renderProtoView);
  }));
}
Object.defineProperty(_collectNestedProtoViewsVariableBindings, "parameters", {get: function() {
    return [[$traceurRuntime.genericType(List, RenderProtoViewWithIndex)]];
  }});
function _createVariableBindings(renderProtoView) {
  var variableBindings = MapWrapper.create();
  MapWrapper.forEach(renderProtoView.variableBindings, (function(mappedName, varName) {
    MapWrapper.set(variableBindings, varName, mappedName);
  }));
  ListWrapper.forEach(renderProtoView.elementBinders, (function(binder) {
    MapWrapper.forEach(binder.variableBindings, (function(mappedName, varName) {
      MapWrapper.set(variableBindings, varName, mappedName);
    }));
  }));
  return variableBindings;
}
function _collectNestedProtoViewsVariableNames(nestedPvsWithIndex, nestedPvVariableBindings) {
  var nestedPvVariableNames = ListWrapper.createFixedSize(nestedPvsWithIndex.length);
  ListWrapper.forEach(nestedPvsWithIndex, (function(pvWithIndex) {
    var parentVariableNames = isPresent(pvWithIndex.parentIndex) ? nestedPvVariableNames[pvWithIndex.parentIndex] : null;
    nestedPvVariableNames[pvWithIndex.index] = _createVariableNames(parentVariableNames, nestedPvVariableBindings[pvWithIndex.index]);
  }));
  return nestedPvVariableNames;
}
Object.defineProperty(_collectNestedProtoViewsVariableNames, "parameters", {get: function() {
    return [[$traceurRuntime.genericType(List, RenderProtoViewWithIndex)], [$traceurRuntime.genericType(List, $traceurRuntime.genericType(Map, $traceurRuntime.type.string, $traceurRuntime.type.string))]];
  }});
function _createVariableNames(parentVariableNames, variableBindings) {
  var variableNames = isPresent(parentVariableNames) ? ListWrapper.clone(parentVariableNames) : [];
  MapWrapper.forEach(variableBindings, (function(local, v) {
    ListWrapper.push(variableNames, local);
  }));
  return variableNames;
}
function _createElementBinders(protoView, elementBinders, allDirectiveBindings) {
  for (var i = 0; i < elementBinders.length; i++) {
    var renderElementBinder = elementBinders[i];
    var dirs = elementBinders[i].directives;
    var parentPeiWithDistance = _findParentProtoElementInjectorWithDistance(i, protoView.elementBinders, elementBinders);
    var directiveBindings = ListWrapper.map(dirs, (function(dir) {
      return allDirectiveBindings[dir.directiveIndex];
    }));
    var componentDirectiveBinding = null;
    if (directiveBindings.length > 0) {
      if (directiveBindings[0].metadata.type === renderApi.DirectiveMetadata.COMPONENT_TYPE) {
        componentDirectiveBinding = directiveBindings[0];
      }
    }
    var protoElementInjector = _createProtoElementInjector(i, parentPeiWithDistance, renderElementBinder, componentDirectiveBinding, directiveBindings);
    _createElementBinder(protoView, i, renderElementBinder, protoElementInjector, componentDirectiveBinding);
  }
}
function _findParentProtoElementInjectorWithDistance(binderIndex, elementBinders, renderElementBinders) {
  var distance = 0;
  do {
    var renderElementBinder = renderElementBinders[binderIndex];
    binderIndex = renderElementBinder.parentIndex;
    if (binderIndex !== -1) {
      distance += renderElementBinder.distanceToParent;
      var elementBinder = elementBinders[binderIndex];
      if (isPresent(elementBinder.protoElementInjector)) {
        return new ParentProtoElementInjectorWithDistance(elementBinder.protoElementInjector, distance);
      }
    }
  } while (binderIndex !== -1);
  return new ParentProtoElementInjectorWithDistance(null, -1);
}
function _createProtoElementInjector(binderIndex, parentPeiWithDistance, renderElementBinder, componentDirectiveBinding, directiveBindings) {
  var protoElementInjector = null;
  var hasVariables = MapWrapper.size(renderElementBinder.variableBindings) > 0;
  if (directiveBindings.length > 0 || hasVariables) {
    protoElementInjector = ProtoElementInjector.create(parentPeiWithDistance.protoElementInjector, binderIndex, directiveBindings, isPresent(componentDirectiveBinding), parentPeiWithDistance.distance);
    protoElementInjector.attributes = renderElementBinder.readAttributes;
    if (hasVariables) {
      protoElementInjector.exportComponent = isPresent(componentDirectiveBinding);
      protoElementInjector.exportElement = isBlank(componentDirectiveBinding);
      var exportImplicitName = MapWrapper.get(renderElementBinder.variableBindings, '\$implicit');
      if (isPresent(exportImplicitName)) {
        protoElementInjector.exportImplicitName = exportImplicitName;
      }
    }
  }
  return protoElementInjector;
}
function _createElementBinder(protoView, boundElementIndex, renderElementBinder, protoElementInjector, componentDirectiveBinding) {
  var parent = null;
  if (renderElementBinder.parentIndex !== -1) {
    parent = protoView.elementBinders[renderElementBinder.parentIndex];
  }
  var elBinder = protoView.bindElement(parent, renderElementBinder.distanceToParent, protoElementInjector, componentDirectiveBinding);
  protoView.bindEvent(renderElementBinder.eventBindings, boundElementIndex, -1);
  MapWrapper.forEach(renderElementBinder.variableBindings, (function(mappedName, varName) {
    MapWrapper.set(protoView.protoLocals, mappedName, null);
  }));
  return elBinder;
}
function _bindDirectiveEvents(protoView, elementBinders) {
  for (var boundElementIndex = 0; boundElementIndex < elementBinders.length; ++boundElementIndex) {
    var dirs = elementBinders[boundElementIndex].directives;
    for (var i = 0; i < dirs.length; i++) {
      var directiveBinder = dirs[i];
      protoView.bindEvent(directiveBinder.eventBindings, boundElementIndex, i);
    }
  }
}
Object.defineProperty(_bindDirectiveEvents, "parameters", {get: function() {
    return [[], [$traceurRuntime.genericType(List, renderApi.ElementBinder)]];
  }});
var RenderProtoViewWithIndex = function RenderProtoViewWithIndex(renderProtoView, index, parentIndex, boundElementIndex) {
  this.renderProtoView = renderProtoView;
  this.index = index;
  this.parentIndex = parentIndex;
  this.boundElementIndex = boundElementIndex;
};
($traceurRuntime.createClass)(RenderProtoViewWithIndex, {}, {});
Object.defineProperty(RenderProtoViewWithIndex, "parameters", {get: function() {
    return [[renderApi.ProtoViewDto], [$traceurRuntime.type.number], [$traceurRuntime.type.number], [$traceurRuntime.type.number]];
  }});
var ParentProtoElementInjectorWithDistance = function ParentProtoElementInjectorWithDistance(protoElementInjector, distance) {
  this.protoElementInjector = protoElementInjector;
  this.distance = distance;
};
($traceurRuntime.createClass)(ParentProtoElementInjectorWithDistance, {}, {});
Object.defineProperty(ParentProtoElementInjectorWithDistance, "parameters", {get: function() {
    return [[ProtoElementInjector], [$traceurRuntime.type.number]];
  }});
