// angular2/src/core/annotations_impl/visibility.js
"use strict";
Object.defineProperties(module.exports, {
  Visibility: {get: function() {
      return Visibility;
    }},
  Self: {get: function() {
      return Self;
    }},
  Parent: {get: function() {
      return Parent;
    }},
  Ancestor: {get: function() {
      return Ancestor;
    }},
  Unbounded: {get: function() {
      return Unbounded;
    }},
  __esModule: {value: true}
});
var $__angular2_47_src_47_facade_47_lang__,
    $__angular2_47_src_47_di_47_annotations_95_impl__;
var CONST = ($__angular2_47_src_47_facade_47_lang__ = require("angular2/src/facade/lang"), $__angular2_47_src_47_facade_47_lang__ && $__angular2_47_src_47_facade_47_lang__.__esModule && $__angular2_47_src_47_facade_47_lang__ || {default: $__angular2_47_src_47_facade_47_lang__}).CONST;
var DependencyAnnotation = ($__angular2_47_src_47_di_47_annotations_95_impl__ = require("angular2/src/di/annotations_impl"), $__angular2_47_src_47_di_47_annotations_95_impl__ && $__angular2_47_src_47_di_47_annotations_95_impl__.__esModule && $__angular2_47_src_47_di_47_annotations_95_impl__ || {default: $__angular2_47_src_47_di_47_annotations_95_impl__}).DependencyAnnotation;
var Visibility = function Visibility(depth, crossComponentBoundaries) {
  $traceurRuntime.superConstructor($Visibility).call(this);
  this.depth = depth;
  this.crossComponentBoundaries = crossComponentBoundaries;
};
var $Visibility = Visibility;
($traceurRuntime.createClass)(Visibility, {shouldIncludeSelf: function() {
    return this.depth === 0;
  }}, {}, DependencyAnnotation);
Object.defineProperty(Visibility, "annotations", {get: function() {
    return [new CONST()];
  }});
Object.defineProperty(Visibility, "parameters", {get: function() {
    return [[$traceurRuntime.type.number], [$traceurRuntime.type.boolean]];
  }});
var Self = function Self() {
  $traceurRuntime.superConstructor($Self).call(this, 0, false);
};
var $Self = Self;
($traceurRuntime.createClass)(Self, {}, {}, Visibility);
Object.defineProperty(Self, "annotations", {get: function() {
    return [new CONST()];
  }});
var Parent = function Parent() {
  $traceurRuntime.superConstructor($Parent).call(this, 1, false);
};
var $Parent = Parent;
($traceurRuntime.createClass)(Parent, {}, {}, Visibility);
Object.defineProperty(Parent, "annotations", {get: function() {
    return [new CONST()];
  }});
var Ancestor = function Ancestor() {
  $traceurRuntime.superConstructor($Ancestor).call(this, 999999, false);
};
var $Ancestor = Ancestor;
($traceurRuntime.createClass)(Ancestor, {}, {}, Visibility);
Object.defineProperty(Ancestor, "annotations", {get: function() {
    return [new CONST()];
  }});
var Unbounded = function Unbounded() {
  $traceurRuntime.superConstructor($Unbounded).call(this, 999999, true);
};
var $Unbounded = Unbounded;
($traceurRuntime.createClass)(Unbounded, {}, {}, Visibility);
Object.defineProperty(Unbounded, "annotations", {get: function() {
    return [new CONST()];
  }});
